### ISSUE TITLE
Rendering lots of objects in nested templates can be really slow

### ISSUE THREAD

**Issue by lekoala** [silverstripe/silverstripe-framework#10174]
## Affected Version

4.x

## Description

I have a project with an editable gridfield with many dropdowns (something like 50 rows and 2 dropdowns per row) and it's super slow to render. Here is a screenshot given by the DebugBar in the admin. It takes about 9 seconds to render all the dropdowns (it's faster on production, but still quite slow).

<img width="1867" alt="Screenshot 2021-12-10 135247" src="https://user-images.githubusercontent.com/250762/145579971-50c991fd-1fcd-4a55-8c69-503fccf84e12.png">
(the CMSTabSet is also super slow, but that's the next issue i'll be handling :) )

In the meantime, i created this
https://github.com/lekoala/silverstripe-base/blob/master/src/Forms/FastDropdownField.php
which is not really a solution, but at least it works for my use case and make the page much faster where i can prerender the template and avoid calling the view engine each time which seem to be the culprit, because it takes less than 1s now.. should be also be there are also a couple of other dropdowns in other tabs).

## Steps to Reproduce

Create an editable gridfield that is not paginated with many dropdowns. Here is some sample code below, simplified for the purposes of this ticket (not tested or isolated on a fresh project).

```
                $c= GridFieldConfig_RecordEditor::create();
                $c->addComponent(new GridFieldTitleHeader());
                $c->removeComponentsByType(GridFieldSortableHeader::class);
                $c->removeComponentsByType(GridFieldPaginator::class);
                $c->removeComponentsByType(GridFieldPageCount::class);
                $c->removeComponentsByType(GridFieldDeleteAction::class);
                $c->removeComponentsByType(GridFieldAddNewButton::class);
                $gf= new GridField('Records', 'Records', $this->owner->List(), $c);
                $fields->addFieldsToTab('Root.Records', $Records);
        
                $c->removeComponentsByType(GridFieldDataColumns::class);
                $c->addComponent($cols = new GridFieldEditableColumns());
               
                $dataMap= $this->dataMap();
                $fields = [];

                $fields['RelationID'] = [
                    'title' => 'Relation',
                    'callback' => function ($r, $name, $grid) use ($dataMap) {
                        $dd = new DropdownField($name, null, $dataMap, $r->$name);
                        return $dd;
                    }
                ];
                $fields['OtherRelationID'] = [
                    'title' => 'Other Relation',
                    'callback' => function ($r, $name, $grid) use ($dataMap) {
                        $dd = new DropdownField($name, null, $dataMap, $r->$name);
                        return $dd;
                    }
                ];
                $cols->setDisplayFields($fields);
```

## Maybe worth exploring

Simplifying the rendering by not using the view engine to render the option list. I'm going to explore that as well (would probably be cleaner than what i did) and update this issue if I found anything

## PRs

[**kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/17026844314)

Note: This PR doesn't fix the problem, it only makes it _slightly_ less of a problem. Don't close the issue after this PR is merged.

- https://github.com/silverstripe/silverstripe-framework/pull/11826

**Comment by lekoala** [comment]
So indeed it seems that using the template slow things down quite a bit and using plain php code is faster at least from what i can tell (it's not exactly easy to compare from within a complex project)

I think it's probably worth dropping the template that is not adding much value nevertheless but maybe more testing needs to be done.

I've updated
https://github.com/lekoala/silverstripe-base/blob/master/src/Forms/FastDropdownField.php
And it's working fine as far as i can tell in my project

**Comment by sb-relaxt-at** [comment]
I had some similar experiences (related to other FormFields) that using a lot of template calls causes serious performance problems.

**Comment by lekoala** [comment]
small update on this because this is becoming really annoying :-)

with a couple of dropdowns with a country list

<img width="856" alt="image" src="https://github.com/silverstripe/silverstripe-framework/assets/250762/8a8b7d1c-422f-497f-98f6-b15aa87c3dc8">

without the dropdown (just went to the cache folder and removed the content on the .dropdown.ss cache file)

<img width="573" alt="image" src="https://github.com/silverstripe/silverstripe-framework/assets/250762/81306109-d490-489a-afe5-cac57ebaf36d">

yes, 3s different for a couple of dropdowns. yes, it's slow due to xdebug profiling the whole thing, but still, i didn't expect such a large difference. i'm going to try to pinpoint where this is coming from.

**Comment by sb-relaxt-at** [comment]
Agreed, we have a component for providing some kind of inline editing by using form fields for editing nested entities. When switching some hidden fields (used for storing internal information like sort order) from templates to plain html strings it resulted in a significant performance improvement. 

From my experience having xdebug enabled can have an extreme performance impact in Silverstripe projects. You might want to try it without, I assume the impact in a production environment will be a lot smaller. Nevertheless I would welcome any performance improvement to the template layer :-)

**Comment by lekoala** [comment]
it seems that every call to $scope->locally() is really slow
which is kind of expected since it does   $this->localStack[] = array_splice($this->itemStack, $this->localIndex + 1); under the hood. so there is a whole lot of array copying happening.

i would qualify this as a design flaw from the template itself. it's quite visible for a dropdown with lots of options, but it's basically affecting all form fields

the long term solution would be: 
- fixing the template language itself (i don't see how that could be easily done)
- use less templates, and avoid templates (or at least, looped over templates) for forms

**Comment by GuySartorelli** [comment]
@lekoala I can't achieve anything like the multiple-seconds-long rendering times you've been seeing with my local tests, so if you can provide some tested reproduction steps I can perform from a fresh installation that would be helpful.
I _suspect_ (though I can't prove without reliable reproduction steps) most of that time would have been getting the source data - I tested with empty `DropdownField` only.

My initial local testing shows that template rendering has sped up a bit in CMS 6, and that most of the difference between a `FastDropdownField`-like implementation and rendering templates is now coming from the new `CastingService` (which is just an abstraction of the logic `ViewableData::obj()` had been doing already). I did have to implement `FastDropdownField::FieldHolder()` to see a consistent difference, though.

Will do some proper profiling over this week to see if I can find any specific hotspots.

**Comment by lekoala** [comment]
things look much worse in dev due to xdebug (and my test project being a bit heavy :) ), but in general, rendering templates is not exactly fast. as long as you only have a couple of .ss and includes, it's not really noticeable, but with hundreds of form items, it gets more noticeable.

i can revisit the issue and see if I can replicate this on a SilverStripe 6 vanilla project

**Comment by GuySartorelli** [comment]
For reference my rough benchmarking was done with the following code in `_config.php`:

```php
<?php

use LeKoala\Base\Forms\FastDropdownField;
use SilverStripe\Control\Controller;
use SilverStripe\Control\HTTPRequest;
use SilverStripe\Control\Session;
use SilverStripe\Forms\DropdownField;
use SilverStripe\Forms\FieldList;
use SilverStripe\Forms\Form;

$fields = [];
foreach (range(0, 30) as $i) {
    $fields[] = new DropdownField('Field' . $i);
    // $fields[] = new FastDropdownField('Field' . $i);
}

Controller::config()->set('url_segment', '123');
$controller = new Controller();
$controller->setRequest((new HTTPRequest('GET', ''))->setSession(new Session([])));

$form = new Form(
    $controller,
    fields: new FieldList($fields)
);
$form->disableSecurityToken();


$start = hrtime(true);
$form->forTemplate();
$end = hrtime(true);

$timeInMs = ($end - $start) / 1e+6;
var_dump("{$timeInMs}ms");
```

The `FastDropdownField` is yours linked above, but with this added implementation of `FieldHolder()` (I skipped the conditional checks for the sake of simplicity):
```php
    public function FieldHolder($properties = [])
    {
        $val = '<div id="' . $this->HolderID() . '" class="field">';
        $val .= '<label class="left" for="' . $this->ID() . '">' . $this->Title() . '</label>';
        $val .= '<div class="middleColumn">';
        $val .= $this->field($properties);
        $val .= '</div></div>';
        return $val;
    }
```

I am getting these results (not on flush - I'm not checking for template parsing time, just template rendering time):

|CMS version|`FastDropdownField`|`DropdownField`|`DropdownField` (skipping `CastingService`)|
|---|---|---|---|
|CMS 5|~18ms|~39ms|N/A|
|CMS 6|~17ms|~23ms|~23ms|

My earlier mention of `CastingService` taking so much time was flawed, running it again today shows that the casting service has nearly no effect on the time it takes to render templates.

**Comment by GuySartorelli** [comment]
> I can revisit the issue and see if I can replicate this on a SilverStripe 6 vanilla project

That would be very helpful if you could. I'll try to do some proper profiling in the meantime.

**Comment by lekoala** [comment]
Ok so here is an update:

On my actual project that had the issue:

37x dropdown field + xdebug **on** : 5.44s (15.6s total)
37x fast dropdown field + xdebug **on** : 0.54s (10.12s total)
37x dropdown field + xdebug **off** : 0.5s (1.59s total)
37x fast dropdown field + xdebug **off**: 0.06s (0.6s total)

it's worth noting that for example checkboxes are not slow, only dropdowns (probably the loop over options)

On a fresh ss6 project

100x dropdown field + xdebug **on** : 115s (118s total) <-- yes!
100x fast dropdown field + xdebug **on** : 0.8s (2.84s total)
100x dropdown field + xdebug **off** : 4.s (4.96s total)
100x fast dropdown field + xdebug **off** : 0.06s (0.3s total)

<img width="1307" height="98" alt="Image" src="https://github.com/user-attachments/assets/3bc329f0-9e17-484d-95b8-89d14b719b17" />
vs
<img width="1965" height="122" alt="Image" src="https://github.com/user-attachments/assets/ca621c45-1310-48c2-86c2-059542cbba06" />

here is the demo code: this is from a ModelAdmin screen with a DemoObject with only one field Title iterating over a list of 100 records (already an array)

```php
  public function getCMSFields()
    {
        $fields = parent::getCMSFields();

        $map = DemoObject::get()->map()->toArray();

        if (count($map) < 100) {
            foreach (range(0, 100) as $i) {
                $o = new DemoObject();
                $o->Title = "Demo $i";
                $o->write();
            }
        }

        foreach (range(1, 100) as $i) {
            // $fields->push(new FastDropdownField("Drop{$i}", null, $map));
            $fields->push(new DropdownField("Drop{$i}", null, $map));
        }

        return $fields;
    }
```

these timings are from the debugbar module, so maybe there is also some overhead due to the proxy/collection

clearly, things go worse when you loop over a large list
and even worse if you have xdebug enabled

**Comment by GuySartorelli** [comment]
Thanks, I can reproduce similar results locally. I'll dive into it and see where I get.

> it's worth noting that for example checkboxes are not slow, only dropdowns (probably the loop over options)

In my local test, replacing `DropdownField` with `CheckboxSetField` gives _worse_ performance, which is note worthy both on its own and in contrast to your findings.

**Comment by GuySartorelli** [comment]
I ran some tests using the same amount of data as you did - 100 dropdowns with 100 sources each. I used raw associative arrays instead of a `DataObject` to cut down on the number of possible causes of performance degradation, and I didn't want any database queries.

I also kept my code in `_config.php` to reduce the amount of code being executed that wasn't directly related to what I wanna check.

I didn't compare against CMS 5 this time since I was able to reproduce similar results to yours.

On average, rendering a form with the above data took **~1100ms** (just over a second).

Here's what I've found so far.

## `ViewLayerData`

### Casting data

Casting data to the relevant object (usually to the relevant `DBField`) takes **~300ms**.

`CastingService` (in constructor): **~200ms**
`CastingService` is also used in ModelData::obj() which is used for hasValue() (ViewLayerData::hasDataValue($name)): **~100ms**

Almost all of that comes from config and injector - I can't see anything about the casting service itself that could be obviously improved.

_Maybe_ we could use singletons instead of getting injector to instantiate new services all the time? If we knew we'd definitely have a `DBField` instance I'd feel more confident about that, but the service intentionally allows for any class that has a `setValue()` method and we can't be 100% sure those will always work well as singletons.

If swapping to singletons we'd also probably have to require a `setName()` method, otherwise the singleton would always have the same name as the first time it got fetched. In an edge case, someone may want to do a `$MyField.Name` (I don't know why they would but it's a thing you can do right now).

Probably the best way to do this would be to check `is_a($serviceName, DBField::class, true)` and only use singletons if that's true - which will be most of the time. 

I did try swapping to singletons to see how it would go, and got a **~200-300ms** improvement.

### Getting raw data

ViewLayerData::getValueFromData(): **~200ms-~300ms**

That's basically `__get()` and `__call()` magic method indirections - we have to try calling methods before getting the property to match expected historical behaviour.

Skipping the dynamic behaviour by adding this at the top of that method shaves off **~100-200ms** on its own:

```php
if (($data instanceof ModelData) && in_array($name, ['Title', 'Selected', 'Disabled', 'Value'])) {
    return $data->getField($name);
}
```

But that requires intimate knowledge of what we're trying to fetch ahead of time.

Interestingly even this shaves off **~50-~100ms**:
```php
if (($data instanceof ModelData)) {
    return $data->obj($name, $arguments);
}
```
We can't add that here though because we want the _raw_ value and `obj()` bakes in the casting.
Skipping `callDataMethod()` explicitly only saves **~50ms** so on its own is probably negliglble.

We can maybe squeak in some `ModelData` specific optimisations here (there are some already so there's precedent for it) - I haven't thought too hard about exactly how that would be implemented. I'd expect such changes to save up to **~100ms** at the scale we're discussing.

### tl;dr

`ViewLayerData` takes roughly half the template render time, but everything it's doing is necessary due to the assumptions baked into Silverstripe CMS Magic(tm), such as using property fetch to get method values and the automatic casting to appropriate `DBField` instances which are configurable and injectable.

It treats all data the same way `ViewableData` used to be explicitly treated, so that you can dump anything you want into a template and have it work the way you expect.

## `ScopeManager`

The array copying for `$itemStack` and other scope array copying in `ScopeManager` is negligible:
Time spent in `locally()` specifically: **~11ms**
Total time spent fiddling with scope/item stack arrays: **~50ms**

I couldn't find any areas of `ScopeManager` that are specifically slow. That said, roughly half the template render time seems to be associated with `ViewLayerData` so presumably the other half should be happening in the scope manager? I can't see where though... my current assumption is that there's actually a larger amount in `ViewLayerData` than I'm seeing, e.g. ~75% with only ~25 in the scope manager.

There are of course little bits being spent before and after the template rendering itself, too.

## Things we can do

- Swapping to using singletons in `CastingService` saves **~2-3ms per dropdown** - if doing only with `DBField` it's _probably_ safe.
- Maybe look at some micro-optimisations for `ModelData` in `ViewLayerData::getValueFromData()` (expect up to **~1ms per dropdown** improvement)

@lekoala if you're able to track down other specific places that can be improved I'll be happy to look into them - for now I'll probably do the above.

I'm also gonna see how the performance compares to twig with equivalent templates at the same scale, both with-and-without `ViewLayerData` in the mix, just as a point of comparison.

**Update:** The singleton option turned out to cause a bunch of problems, so we're not doing that.

**Comment by sb-relaxt-at** [comment]
Thank you for your extensive tests!

> I'm also gonna see how the performance compares to twig with equivalent templates at the same scale, both with-and-without `ViewLayerData` in the mix, just as a point of comparison.

That would be very interesting. From my performance struggles in Silverstripe 4, I thought the template layer to be the bottleneck. But my testing was less thorough than yours so it may very well have been the casting (Injector/Config) in the first place. 

I am wondering why the Injector/Config takes so long. As this is a central component, any improvement there could be very benefical.

**Comment by lekoala** [comment]
@GuySartorelli thanks for your findings!

here is a couple of things I can add on my end

There are two culprits for the situation:

- the actual rendering of the option tag : as you discovered, outputting Disabled, Selected, Value.XML and Title.XML + exists check is slow (it shouldn't, but due to how the engine works that seems hard to avoid while keeping expectations)

I did this which works great and seems like a possible compromise (keeping the template, but create the option tag in html)

```html
<select $AttributesHTML>
<% loop $Options %>
	$OptionHTML.RAW
<% end_loop %>
</select>
```

```php
    protected function getFieldOption($value, $title)
    {
        // Check selection
        $selected = $this->isSelectedValue($value, $this->getValue());

        // Check disabled
        $disabled = false;
        if ($this->isDisabledValue($value) && $title != $this->getEmptyString()) {
            $disabled = 'disabled';
        }

        $selectedHTML = $selected ? ' selected="selected"' : '';
        $disabledHTML = $disabled ? ' disabled' : '';
        $titleHTML = strlen($title) > 0 ? Convert::raw2xml($title) : '&nbsp;';
        $valueHTML = Convert::raw2xml($value);
        $item = '<option value="' . $valueHTML . '"' . $selectedHTML . $disabledHTML . '>' . $titleHTML . '</option>';

        return new ArrayData([
            'Title' => (string)$title,
            'Value' => $value,
            'Selected' => $selected,
            'Disabled' => $disabled,
            'OptionHTML' => $item,
        ]);
    }
```

This is two times faster than rendering the option with the template engine

- the loop itself is slow due to context switching and object lookup

so we might as well do this

```html
<select $AttributesHTML>
    $OptionsHTML
</select>
```

at least there is a minimal template to keep compatibility (instead of what I do in my FastDropdownField class) but the loop is avoided entirely

- i don't see any obvious or easy way to improve how the template engine works. I think if there was some kind of "simple loops" that don't allow context switching (no Up, Top...) that would be faster, typically : instead of having the ->locally() / ->resetScope() for each item in the loop, it would get the list once, then loop over it and call the various fields on the list ($Title, $Value etc would call $item->Title, $item->Value...) but that seems an entirely different issue / feature

**Comment by GuySartorelli** [comment]
I would recommend having the options HTML built in PHP be a project-level customisation rather than in core. You may want to do the same with other form fields that subclass the `SelectField` class.

>  I think if there was some kind of "simple loops" that don't allow context switching (no Up, Top...) that would be faster, typically : instead of having the `->locally() / ->resetScope()` for each item in the loop, it would get the list once, then loop over it and call the various fields on the list (`$Title`, `$Value` etc would call `$item->Title`, `$item->Value`...) but that seems an entirely different issue / feature

As noted in my findings, all of the item scope stuff _altogether_ only takes up ~50ms in this extreme example - I'd expect probably less than half that in improvements from adding a simplified loop which I don't think would be worth the added complexity.

**Comment by GuySartorelli** [comment]
> I am wondering why the Injector/Config takes so long. As this is a central component, any improvement there could be very benefical.

Bare in mind we're talking about a couple of milliseconds per dropdownfield, which is gonna be one dependency injection for each piece of data we want, so 4-or-so for each option - for 100 options. So ~0.02ms per injection (a rough estimate but it demonstrates the idea).

**Comment by GuySartorelli** [comment]
>> I'm also gonna see how the performance compares to twig with equivalent templates at the same scale, both with-and-without ViewLayerData in the mix, just as a point of comparison.
>
> That would be very interesting. From my performance struggles in Silverstripe 4, I thought the template layer to be the bottleneck. But my testing was less thorough than yours so it may very well have been the casting (Injector/Config) in the first place.

My results (using the same data as previous tests) were as follows:

||SS renderer|twig renderer|
|---|---|---|
|With `ViewLayerData`|~1050ms|~1000ms|
|Without `ViewLayerData`|~350ms|~300ms|

The margin of error is pretty big - I was seeing pretty big fluctuations in some runs, so this is averaged out (applies to all my other findings too).

It's pretty clear that _most_ of the render time is due to `ViewLayerData`, `CastingService`, etc - i.e. actually fetching the data in a way that upholds the "magic" of SS templates.

The actual template renderer itself is roughly on par with twig. The main reason twig will seem faster to people who use it in other projects is likely due to the fact that its casting is all done in the view layer, while we do it in the model layer.
That difference seems conceptually small but it ultimately means we have to allow for dynamic functionality - while twig expects you to explicitly cast inside the template itself and therefore doesn't have to manage all those edge cases and dynamic magic nonsense.

That ends up being basically the same difference between hardcoded HTML in `DropdownField::Field()` and rendering a template - the template and view layers have to allow for a lot of dynamic stuff whereas with hardcoded HTML markup you can skip straight to just getting the data you need in the way you need it.

I could see a future where we decide we don't want all that dynamic casting, and move to a system more consistent with twig. I can also see a future where we make the `ViewLayerData` (and corresponding dynamic casting) optional, so large scale projects can do their casting in the template instead with the trade off of performance vs mental overhead.
Any such changes would need to be done in a major release and carefully considered.

**Comment by GuySartorelli** [comment]
I've changed the issue title to reflect what the focus of discussion has been, so it'll be easier to find in the future

**Comment by lekoala** [comment]
> I could see a future where we decide we don't want all that dynamic casting, and move to a system more consistent with twig. I can also see a future where we make the `ViewLayerData` (and corresponding dynamic casting) optional, so large scale projects can do their casting in the template instead with the trade off of performance vs mental overhead. Any such changes would need to be done in a major release and carefully considered.

Also, while it's great and flexible to have all this casting done, in some cases, it's really not needed because as a developer, you know it is a safe variable. Using .RAW should be more or less equivalent than just getting the property on the object without any magic applied.

so all things considered, this:

```
<% loop MyList %>
Title : $Title.RAW
<% end_loop %>
```

should roughly (scoping aside) match this

```
$html  = '';
foreach($list as $item) {
   $html .= 'Title : ' . $item->Title;
}
```

and using `$HTML.RAW` in the template

or at least, we should have some features in the template engine that allows us to skip the magic and use simpler things

**Comment by GuySartorelli** [comment]
> we should have some features in the template engine that allows us to skip the magic and use simpler things

I'd be open to a proposal of how that would work in a way that only affects the templates it's meant to affect without being an additional burden having to remember when to use it and when not to.
A global setting would require developers to replace a bunch of templates from `silverstripe/admin`, `silverstripe/cms` etc.

My current thinking is it would be on a per-model and/or per-controller basis (the same places you can swap the whole template engine out)

**Comment by emteknetnz** [comment]
Have merged PR, will be released with CMS 6.1. Have assigned back to Guy for any potential follow up.
